src/librustc_mir/interpret/eval_context.rs

```diff
@@ -252,17 +252,22 @@ impl<'a, 'mir, 'tcx: 'mir, M: Machine<'a, 'mir, 'tcx>> InterpretCx<'a, 'mir, 'tc
         &self,
         substs: T,
     ) -> EvalResult<'tcx, T> {
+        if !substs.needs_subst() {
+            return Ok(self.tcx.erase_regions(&substs));
+        }
+        // Stack frames beyond the first stackframe have their own set of substs and it would be
+        // absolutely wrong to use them for any further substitutions. So we just yield the substs
+        // after region erasing and let the caller figure out if that is good enough for them.
```

Comment:
I'm confused, why is *the thing to substitute* being called `substs`? Usually we call it something generic like `value`.
Also, this doesn't seem right.

What I would do, if there is any stack frame, is always call `subst_and_normalize_erasing_regions` and *then* check if that still `needs_subst` - because if it does, *the entire* execution (i.e. what `self.param_env` applies to) is "too generic".

Reply:
Consider the following function

```rust
fn foo<U>() -> usize {
    let v: Vec<U> = Vec::new();
    v.len()
}
```

When evaluating `Vec::new()` e.g. due to const propagation, we'd create a stack frame for `Vec::new`. If the type were `Vec<i32>` the substs would be `[i32]` and we'd substitute and be monomorphic from now on. Since the type is `Vec<U>`, we'd substitute the `T` in `Vec<T>::new` with `U` (our substs are `[U]`). This is why we need to substitute the `[T]` substs of `Vec<T>::new` with `[U]`.

We can preemptively return `TooGeneric` as you suggested, but that would prevent us from const propagating calls to `foo`, even though that is totally fine to evaluate even though it's still polymorphic.

Reply:
Okay then you should just never return `TooGeneric`, even when there are no frames?

And in your example, with the code as written, you would have `[T]` and not substitute it to/with `[U]` because you're just not calling `subst_and_normalize_erasing_regions` anymore in that case.
Am I missing something?